airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/spec/SpecTest.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.test.spec
+
+import com.deblock.jsondiff.DiffGenerator
+import com.deblock.jsondiff.diff.JsonDiff
+import com.deblock.jsondiff.matcher.CompositeJsonMatcher
+import com.deblock.jsondiff.matcher.JsonMatcher
+import com.deblock.jsondiff.matcher.LenientJsonObjectPartialMatcher
+import com.deblock.jsondiff.matcher.StrictJsonArrayPartialMatcher
+import com.deblock.jsondiff.matcher.StrictPrimitivePartialMatcher
+import com.deblock.jsondiff.viewer.OnlyErrorDiffViewer
+import com.fasterxml.jackson.databind.ObjectMapper
+import io.airbyte.cdk.test.util.DestinationProcessFactory
+import io.airbyte.cdk.test.util.FakeDataDumper
+import io.airbyte.cdk.test.util.IntegrationTest
+import io.airbyte.cdk.test.util.NoopDestinationCleaner
+import io.airbyte.cdk.test.util.NoopExpectedRecordMapper
+import io.airbyte.protocol.models.Jsons
+import io.airbyte.protocol.models.v0.AirbyteMessage
+import java.nio.file.Files
+import java.nio.file.Path
+import kotlin.test.assertEquals
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Test
+
+private const val EXPECTED_SPEC_FILENAME = "expected_spec.json"
+
+/**
+ * This is largely copied from [io.airbyte.cdk.spec.SpecTest], but adapted to use our
+ * [DestinationProcessFactory].
+ *
+ * It also automatically writes the actual spec back to `expected_spec.json` for easier inspection
+ * of the diff. This diff is _really messy_ for the initial migration from the old CDK to the new
+ * one, but after that, it should be pretty readable.
+ */
+open abstract class SpecTest :
+    IntegrationTest(
+        FakeDataDumper,
+        NoopDestinationCleaner,
+        NoopExpectedRecordMapper,
+    ) {
+    @Test
+    fun testSpec() {
```

Comment:
Can we please stop this pattern of defining junit tests in abstract classes? They're like a straightjacket and I strongly believe this pattern is responsible for the poor quality of the testing of the legacy connectors.

Instead, I encourage you to create test fixtures and define one-liner junit tests in the connectors.

Composition over inheritance blah blah blah...

Reply:
(tl;dr imo the abstract class thing is the least-shitty pattern available)

hm, are you suggesting something like this? (also can you say more about what goes wrong with the abstract class pattern? I personally think it's... sufficiently not terrible, though obviously it's pretty sucky)
```kotlin
package cdk.blah
class DestinationTests {
  fun testSpec() {...}
  fun testWrite() {...}
}

package destination-blah
class DestinationBlahSpecTest {
  @Test
  fun testSpec() { DestinationTests.testSpec() }

  @Test
  fun testWrite() { DestinationTests.testWrite() }
}
```

if I'm reading you correctly, then I disagree. Specifically: I think these two requirements are nonnegotiable:
* test cases must be opt-out (i.e. if we add a new test case in CDK v1.2.3, then upgrading destination-foo to cdk 1.2.3 should automatically activate that test for destination-foo)
    * (reasoning: nobody is going to read through the cdk changelog and remember to opt into new tests, every time they bump the cdk version. And destinations want to put most of our tests in the CDK.)
* it must be easy to run an individual test case from intellij (it sucks that this is a real problem, but de facto we all use intellij, and it imposes some annoying constraints on the problem - e.g. we can't use DynamicTest, parameterized tests are a little annoying, etc).
    * (IMO it's self-evident that this is important)

and requiring destination authors to write those `fun testFoo() { Base.testFoo() }` methods fails the first criterion

Reply:
haven't done any research myself (obviously kotest looms large, with our shift to kotlin, but I know nothing about the pros/cons). I think the problem is more with intellij than the test framework though (with 0 evidence)

Reply:
also, I have a note to myself - this test (unlike most of our other tests) actually _could_ live purely in the CDK as a concrete class. Just need to get gradle/junit to search the full classpath, instead of just the destination-foo jar. Currently `gw :air-int:conn:dest-e2e-test:integrationTestJava --tests io.airbyte.cdk.test.spec.SpecTest` errors b/c no tests are discovered, but that feels solvable.

(or something along those lines, I didn't really dig into it)